### PR TITLE
Area measurment Mesh Generation

### PR DESCRIPTION
**Pull request by Jossos**
- Implemented Area Measurement Mesh Generator
- Fixes [AB#1980](https://dev.azure.com/euclideon/57c3bd2a-8f94-4578-b195-2e4fa9d8f295/_workitems/edit/1980)
- Fixes [AB#1769](https://dev.azure.com/euclideon/57c3bd2a-8f94-4578-b195-2e4fa9d8f295/_workitems/edit/1769)
- Fixes [AB#2116](https://dev.azure.com/euclideon/57c3bd2a-8f94-4578-b195-2e4fa9d8f295/_workitems/edit/2116)

### CODE REVIEW

File: src/scene/vcPOI.cpp [hunk 1]
```diff
@@ -1331,69 +1332,512 @@ void vcPOI::AddFillPolygonToScene(vcState *pProgramState, vcRenderData *pRenderD
...
+            {
+              if ((int)i != pointConnection2 && pointConnection != pointConnection2)
+              {
+                udInt2 line2 = udInt2::create((int32_t)i2, (int32_t)pointConnection2);
+
+                udDouble3 splitPoint;
+                if (vcPOI_LineLineIntersection2D(pointPositions[line1.x], pointPositions[line1.y], pointPositions[line2.x], pointPositions[line2.y], &splitPoint))
+                {
+                  // New point to connect the 4 other points
+                  udDouble3 newPointPosition;
```
Comment (peterAdams1014) [inline src/scene/vcPOI.cpp:1473]: `newPointConnection` can be removed.

File: src/scene/vcPOI.cpp [hunk 2]
```diff
@@ -1331,69 +1332,512 @@ void vcPOI::AddFillPolygonToScene(vcState *pProgramState, vcRenderData *pRenderD
...
+
+  int prevPoint = startPoint;
+  int currentPoint = startPoint;
+  double prevClosestAngle = DBL_MAX;
+  for (const int &pointIndex : pointConnections[startPoint])
+  {
+    const udDouble3 &pos = pointPositions[startPoint];
+    const udDouble3 &target = pointPositions[pointIndex];
+
+    double angle = -atan2(pos.y - target.y, pos.x - target.x);
```
Comment (peterAdams1014) [inline src/scene/vcPOI.cpp:1542]: `udAtan2()`

File: src/scene/vcPOI.cpp [hunk 3]
```diff
@@ -1331,69 +1332,512 @@ void vcPOI::AddFillPolygonToScene(vcState *pProgramState, vcRenderData *pRenderD
...
+    for (size_t i = 0; i < pointInfoPositions.length; ++i)
+    {
+      int prevIndex = (int)(i == 0 ? (pointInfoPositions.length - 1) : i - 1);
+      int nextIndex = (int)(i == (pointInfoPositions.length - 1) ? 0 : i + 1);
+
+      udDouble2 curPointPos2 = udDouble2::create(pointInfoPositions[i].x, pointInfoPositions[i].y);
+      udDouble2 nextPointPos2 = udDouble2::create(pointInfoPositions[nextIndex].x, pointInfoPositions[nextIndex].y);
+      udDouble2 prevPointPos2 = udDouble2::create(pointInfoPositions[prevIndex].x, pointInfoPositions[prevIndex].y);
+
+      if (udMag2(curPointPos2 - prevPointPos2) < UD_EPSILON
```
Comment (peterAdams1014) [inline src/scene/vcPOI.cpp:1634]: Not sure how often this runs, but potential optimization could be done here:

`udMagSq2() < UD_EPSILON * UD_EPSILON`

File: src/scene/vcPOI.cpp [hunk 4]
```diff
@@ -1331,69 +1332,512 @@ void vcPOI::AddFillPolygonToScene(vcState *pProgramState, vcRenderData *pRenderD
...
+  for (size_t i = 0; i < pTrianglePoints->length; ++i)
+  {
+    double closestDist = FLT_MAX;
+    double closestZ = 0;
+    for (const udDouble3 &point : perimeterPoints)
+    {
+      udDouble2 triPoint2 = udDouble2::create(pTrianglePoints->GetElement(i)->x, pTrianglePoints->GetElement(i)->y);
+      udDouble2 pos = (point - pivotPoint).toVector2();
+
+      double dist = udMag2<double>(pos - triPoint2);
```
Comment (peterAdams1014) [inline src/scene/vcPOI.cpp:1756]: Compare squared values here.

**Review by peterAdams1014 — CHANGES_REQUESTED** [review]

File: src/scene/vcPOI.cpp (hunk 1 above)
Comment (Jossos) [reply src/scene/vcPOI.cpp:1473]: newPointPosition*

**Review by peterAdams1014 — APPROVED** [review]

**Pull request merged**